### PR TITLE
feat: introduce `UdpSocketState::try_send` APIs

### PR DESCRIPTION
**Pull request by thomaseizinger**
Resolves: #1971.

### CODE REVIEW

File: quinn-udp/src/unix.rs [hunk 1]
```diff
@@ -191,7 +191,21 @@ impl UdpSocketState {
...
+        match send(self, socket.0, transmit) {
+            Ok(()) => Ok(()),
+            Err(e) if e.kind() == io::ErrorKind::WouldBlock => Err(e),
+            Err(e) => {
+                // Other errors are ignored, since they will usually be handled
+                // by higher level retransmits and timeouts.
+                // - PermissionDenied errors have been observed due to iptable rules.
+                //   Those are not fatal errors, since the
+                //   configuration can be dynamically changed.
+                // - Destination unreachable errors have been observed for other
```
Comment (djc) [inline quinn-udp/src/unix.rs:203]: Feels like this should be a docstring on the public API?

File: quinn-udp/src/windows.rs [hunk 2]
```diff
@@ -358,7 +360,13 @@ fn send(socket: UdpSockRef<'_>, transmit: &Transmit<'_>) -> i32 {
             ptr::null_mut(),
             None,
         )
+    };
+
+    if rc != 0 {
+        return Err(io::Error::last_os_error());
     }
+
+    Ok(())
```
Comment (djc) [inline quinn-udp/src/windows.rs]: Nit: prefer spelling this like

```rust
match rc {
     0 => Ok(()),
     _ => Err(io::Error::last_os_error()),
}
```

**Review by djc — APPROVED** [review]
This looks reasonable to me, thanks!

**Comment by thomaseizinger** [comment]
@djc Can the feedback be in follow-up commits or should I append?

**Comment by djc** [comment]
> @djc Can the feedback be in follow-up commits or should I amend?

Please amend so that we end up with a clean commit series to review.

**Comment by thomaseizinger** [comment]
Just fixed up CI, sorry for the hickup!

**Review by Ralith — APPROVED** [review]
Thanks! We should probably consolidate the public definitions somehow one of these days...

**Pull request merged**